File: py800mon/ui/inputwidget.py
import curses

from ..app import VisualComponent
from .color import Color


class InputWidget(VisualComponent):
    def __init__(
        self,
        window,
        color: Color = Color.TEXT,
        max_length: int | None = None,
        on_enter=None,
        on_change=None,
    ):
        super().__init__(window)
        self.color = color
        self.max_length = max_length
        self.on_enter = on_enter
        self.on_change = on_change
        self.invalid = False
        self._buffer = ""

    @property
    def buffer(self) -> str:
        return self._buffer

    def activate(self, initial_buffer: str = ""):
        text = str(initial_buffer)
        if self.max_length is not None:
            text = text[: self.max_length]
        self._buffer = text

    def deactivate(self):
        self.invalid = False
        self._buffer = ""

    def set_invalid(self, invalid: bool):
        self.invalid = invalid

    def _normalize_char(self, ch: str) -> str:
        return ch

    def _is_char_allowed(self, ch: str) -> bool:
        code = ord(ch)
        return 32 <= code <= 126

    def _to_value(self):
        return self._buffer

    def set_buffer(self, value: str) -> None:
        text = str(value)
        if self.max_length is not None:
            text = text[: self.max_length]
        if text == self._buffer:
            return
        self._buffer = text
        self.emit_change()

    def _append_char(self, ch: int) -> None:
        if ch < 0 or ch > 255:
            return
        char = self._normalize_char(chr(ch))
        if not self._is_char_allowed(char):
            return
        if self.max_length is not None and len(self._buffer) >= self.max_length:
            return
        self._buffer += char
        self.emit_change()

    def _backspace(self) -> None:
        if not self._buffer:
            return
        self._buffer = self._buffer[:-1]
        self.emit_change()

    def handle_key(self, ch: int) -> bool:
        if ch in (curses.KEY_BACKSPACE, 127, 8):
            self._backspace()
            return True
        before = self._buffer
        self._append_char(ch)
        return self._buffer != before

    def emit_change(self):
        if not self.on_change:
            return
        self.on_change(self._to_value())

    def emit_enter(self):
        if not self.on_enter:
            return
        self.on_enter(self._to_value())

    def render(self, force_redraw=False):
        self.window.cursor = 0, 0
        color = Color.INPUT_INVALID if self.invalid else self.color
        attr = color.attr() | curses.A_REVERSE
        text = self._buffer
        self.window.print(text, attr=attr)
        self.window.fill_to_eol(attr=attr)
        cursor_x = len(text)
        if cursor_x > self.window._iw - 1:
            cursor_x = self.window._iw - 1
        if cursor_x < 0:
            cursor_x = 0
        self.window.cursor = (cursor_x, 0)

File: go800mon/a800mon/monitor/inputwidget.go
package monitor

import (
	"strings"

	. "go800mon/a800mon"
)

type InputWidget struct {
	BaseWindowComponent
	color     Color
	buffer    string
	maxLength int
	onChange  func(string)
	invalid   bool
	normalize func(rune) rune
	validate  func(rune) bool
}

func NewInputWidget(window *Window) *InputWidget {
	return &InputWidget{
		BaseWindowComponent: NewBaseWindowComponent(window),
		color:               ColorText,
		maxLength:           -1,
		normalize:           normalizeRuneIdentity,
		validate:            validatePrintableASCII,
	}
}

func (i *InputWidget) Render(_force bool) {
	w := i.Window()
	w.Cursor(0, 0)
	color := i.color
	if i.invalid {
		color = ColorInputInvalid
	}
	attr := color.Attr() | AttrReverse()
	text := i.buffer
	w.Print(text, attr, false)
	w.FillToEOL(' ', attr)
	cursorX := len([]rune(text))
	width := w.Width()
	if width <= 0 {
		cursorX = 0
	} else if cursorX > width-1 {
		cursorX = width - 1
	}
	if cursorX < 0 {
		cursorX = 0
	}
	w.Cursor(cursorX, 0)
}

func (i *InputWidget) SetColor(color Color) {
	i.color = color
}

func (i *InputWidget) SetMaxLength(n int) {
	if n <= 0 {
		i.maxLength = -1
		return
	}
	i.maxLength = n
}

func (i *InputWidget) SetOnChange(cb func(string)) {
	i.onChange = cb
}

func (i *InputWidget) SetCharNormalizer(normalize func(rune) rune) {
	if normalize == nil {
		i.normalize = normalizeRuneIdentity
		return
	}
	i.normalize = normalize
}

func (i *InputWidget) SetCharValidator(validate func(rune) bool) {
	if validate == nil {
		i.validate = validatePrintableASCII
		return
	}
	i.validate = validate
}

func (i *InputWidget) AcceptsChar(ch int) bool {
	_, ok := i.normalizeChecked(ch)
	return ok
}

func (i *InputWidget) Buffer() string {
	return i.buffer
}

func (i *InputWidget) Activate(initial string) {
	i.SetBuffer(initial)
}

func (i *InputWidget) Deactivate() {
	i.invalid = false
	i.SetBuffer("")
}

func (i *InputWidget) SetBuffer(text string) {
	out := []rune(text)
	if i.maxLength > 0 && len(out) > i.maxLength {
		out = out[:i.maxLength]
	}
	normalized := string(out)
	if normalized == i.buffer {
		return
	}
	i.buffer = normalized
	i.emitChange()
}

func (i *InputWidget) backspace() {
	r := []rune(i.buffer)
	if len(r) == 0 {
		return
	}
	i.buffer = string(r[:len(r)-1])
	i.emitChange()
}

func (i *InputWidget) HandleKey(ch int) bool {
	if ch == KeyBackspace() || ch == 127 || ch == 8 {
		i.backspace()
		return true
	}
	r, ok := i.normalizeChecked(ch)
	if !ok {
		return false
	}
	if i.maxLength > 0 && len([]rune(i.buffer)) >= i.maxLength {
		return false
	}
	i.buffer += string(r)
	i.emitChange()
	return true
}

func (i *InputWidget) SetInvalid(invalid bool) {
	i.invalid = invalid
}

func (i *InputWidget) Invalid() bool {
	return i.invalid
}

func (i *InputWidget) emitChange() {
	if i.onChange == nil {
		return
	}
	i.onChange(i.buffer)
}

func (i *InputWidget) normalizeChecked(ch int) (rune, bool) {
	if ch < 0 || ch > 255 {
		return 0, false
	}
	r := i.normalize(rune(ch))
	if !i.validate(r) {
		return 0, false
	}
	return r, true
}

type AddressInputWidget struct {
	*InputWidget
}

func NewAddressInputWidget(window *Window) *AddressInputWidget {
	w := NewInputWidget(window)
	w.SetMaxLength(4)
	w.SetCharNormalizer(func(r rune) rune {
		if r >= 'a' && r <= 'z' {
			return r - 32
		}
		return r
	})
	w.SetCharValidator(func(r rune) bool {
		return (r >= '0' && r <= '9') || (r >= 'A' && r <= 'F')
	})
	return &AddressInputWidget{InputWidget: w}
}

func (a *AddressInputWidget) Render(_force bool) {
	w := a.Window()
	w.Cursor(0, 0)
	color := a.color
	if a.invalid {
		color = ColorInputInvalid
	}
	attr := color.Attr() | AttrReverse()
	text := addressInputDisplayText(a.buffer)
	w.Print(text, attr, false)
	w.FillToEOL(' ', attr)
	w.Cursor(4, 0)
}

func normalizeRuneIdentity(r rune) rune {
	return r
}

func validatePrintableASCII(r rune) bool {
	return r >= 32 && r <= 126
}

func addressInputDisplayText(text string) string {
	upper := strings.ToUpper(text)
	runes := []rune(upper)
	if len(runes) > 4 {
		runes = runes[len(runes)-4:]
	}
	for len(runes) < 4 {
		runes = append([]rune{'0'}, runes...)
	}
	return string(runes)
}
